### ISSUE TITLE
Notifications do not propagate to an AppDelegate event handler

### ISSUE THREAD

**Issue by tones111** [linebender/druid#1879]
I'm just starting to experiment with druid and ran into some confusion when trying to handle notifications within an AppDelegate event function.  It appears the top-level window eats it after printing out a couple INFO logs.

After running into this issue I came across #1388 that lists a use case of implementing list selection via a notification.  I'm attempting to have a Label widget from my list notify my AppDelegate that it's been selected.  The delegate will then send commands to the label to change it's background color (and tell any previously selected label to unset its background).  Eventually there will be more orchestration commands to different sections of UI.

The AppDelegate seems like a good place to put this orchestration logic.  Is this lack of propagation intentional or an oversight?

Thanks for all the effort going into druid.  I really like the composability of the various widgets and it feels well designed!  Looking forward to watching it mature.

**Comment by lisael** [comment]
From my understanding, an AppDelegate is somewhat "off-tree". It exists to give the user a chance to modify an event that pass through it before hitting the App, the real root of the widget tree.

What you need is a `ControllerHost<W, C>` (a widget that wraps another one and delegates calls to `event`, `lifecycle` and `update` to a user-defined `Controller<T, W: Widget<T>>`) at the top of your widget hierarchy.

See https://github.com/linebender/druid/blob/master/druid/src/widget/controller.rs

**Comment by tones111** [comment]
Thanks for the insight.  I think I had the impression commands were primarily meant for going down the tree while notifications go up.  As you mentioned, since the command is targeted it makes sense that there's much less overhead to deliver.  Definitely some good ideas to work with here.  Thanks!

**Comment by lisael** [comment]
Technically speaking, commands still crawl all tree from top to bottom, as internal events. It's the wrapping `WidgetPod`s that do forward them to their respective inner widget as plain `Event::Command`s if the inner widget is the target (see `WidgetPod::event(...)` for details). Preferring commands over  notifications when applicable is more a matter of code cleanness that performance.

There probably is a performance impact, though as in case of a notification the event() method on inner widget is called whatsoever.

**Comment by lisael** [comment]
Also. Please take all I say with a grain of salt, I'm quite new to Druid, and there are probably some subtleties I don't know and I may have wrong assumptions about the internals. I think there is a (well known) lack of documentation, but the code of the internals is clear and easy to follow IMHO. I head there when I feel stuck, it helps a lot.

**Comment by cmyr** [comment]
This is correct, the AppDelegate is not part of the tree. To be honest, I hadn't even thought about the AppDelegate when I implemented notifications, and it isn't unreasonable that they could be passed there, but it really wasn't the design goal: notifications are for communicating state between a child and parent widget, and a command is a better fit if you want to get something to your `AppDelegate`.
